File: mqtt_chat_ruby/chat_client.rb
#!/usr/bin/env ruby 
# encoding: utf-8
#
# MQTT Chat Client
#
# Author : Maho Takara  takara@jp.ibm.com
# 2016/11/26  Initial commit
#

require 'mqtt'
require 'json'

##### MAIN #####
if __FILE__ == $0

  cnf = open("config.json") do |io|
    JSON.load(io)
  end

  print "My reply topic ", cnf['reply_topic'], "\n"

  client = MQTT::Client.connect(cnf['mqtt_broker'])
  client.subscribe( cnf['reply_topic'] )

  while true
    print "> "
    text = gets.chop

    hash_message = {}
    hash_message['text'] = text
    hash_message['reply'] = cnf['reply_topic']
    json_message = JSON.generate(hash_message)

    client.publish( cnf['listen_topic'], json_message, retain=false)

    topic,message = client.get
    puts "#{message}"
  end

  client.disconnect()
end

File: mqtt_chat_node/chat_client.js
#!/usr/bin/env node
//
// MQTT Chat Client
//
// Author : Maho Takara  takara@jp.ibm.com
// 2016/11/26  Initial commit
//

var mqtt    = require('mqtt');
var cnf = require('./config.json');
var client  = mqtt.connect(cnf.broker_url);

// Callback as the connection established
client.on('connect', function () {
    console.log("connect & subscribe");
    client.subscribe(cnf.reply_topic);
});

// Callbask as the message received
client.on('message', function (topic, message) {
    console.log("=== receive message ===");
    console.log("topic: ", topic.toString());
    console.log("message: ", message.toString());
});

// Setup keyboard input
process.stdin.resume();
process.stdin.setEncoding('utf8');
var util = require('util');

// Callback when enterkey is pushed
process.stdin.on('data', function (text) {
    text = text.slice(0, -1);  // chop
    console.log('key input data:', util.inspect(text));
    json_message = { "reply": cnf.reply_topic, "text": text};
    text_message = JSON.stringify(json_message, null, 0);
    client.publish(cnf.broker_topic, text_message);
    if (text === 'bye') {
	done();
    }
});

// Process quit
function done() {
    console.log('end');
    client.end();
    process.exit();
}
